src/main/java/eu/europeana/metis/sandbox/service/workflow/harvest/FileHarvestService.java

```diff
@@ -76,40 +77,41 @@ private HarvestIdentifiersResult<String> harvestIdentifiersFromCompressedArchive
       throws ServiceException {
     final int numberOfRecordsToStepInto = normalizeStepSize(stepSize);
     final List<String> result = new ArrayList<>();
-    int recordCounter = 0;
-    int skipCounter = 0;
-    boolean recordLimitExceeded;
+    int currentIndex = 0;
+    int nextIndexToSelect = numberOfRecordsToStepInto - 1;
 
     try (InputStream bis = new ByteArrayInputStream(fileContent);
         BufferedInputStream buffered = new BufferedInputStream(bis);
         ArchiveInputStream<?> ais = new ArchiveStreamFactory().createArchiveInputStream(buffered)) {
 
       ArchiveEntry entry;
-      while ((entry = ais.getNextEntry()) != null && recordCounter < maxAllowedRecords) {
+      while ((entry = ais.getNextEntry()) != null && result.size() < maxAllowedRecords) {
         if (entry.isDirectory() || shouldSkip(entry.getName())) {
           continue;
         }
 
-        if (skipCounter == 0) {
+        if (currentIndex == nextIndexToSelect) {
           log.debug("Processing entry {}", entry.getName());
           result.add(entry.getName());
-          recordCounter++;
-          skipCounter = numberOfRecordsToStepInto - 1;
-        } else {
-          skipCounter--;
+          nextIndexToSelect += numberOfRecordsToStepInto;
         }
+        currentIndex++;
       }
 
-      recordLimitExceeded = recordCounter >= maxAllowedRecords;
-      if (recordLimitExceeded) {
-        log.warn("Reached maximum number of records ({}) during harvesting.", maxAllowedRecords);
+      if (isStepSizeBiggerThanDatasetSize(result.size(), currentIndex)) {
+        throw new StepIsTooBigException(currentIndex);
       }
     } catch (IOException e) {
       throw new ServiceException("Error harvesting records ", e);
     }
+    boolean recordLimitExceeded = result.size() >= maxAllowedRecords;
```

Comment:
should be this `result.size() > maxAllowedRecords;` ?
thinking on the upper boundary case.
before 
```
1000 >=1000    <=this should not be flagged as recordLimitExceeded
1001 >=1000    <=this is fine
```
after
```
1000 > 1000 => false
1001 > 1000 => true
```
what do you think?

Reply:
The index starts from zero.

Reply:
:sweat_smile: true. got it.